File: DynamicKey/AgoraDynamicKey/ruby/lib/dynamic_key2/education_token_builder.rb
module AgoraDynamicKey2
  class EducationTokenBuilder
    # Build room user token.
    #
    # app_id: The App ID issued to you by Agora. Apply for a new App ID from Agora Dashboard if it is missing
    #     from your kit. See Get an App ID.
    # app_certificate: Certificate of the application that you registered in the Agora Dashboard.
    #     See Get an App Certificate.
    # room_uuid: The room's id, must be unique.
    # user_uuid: The user's id, must be unique.
    # role: The user's role.
    # expire: represented by the number of seconds elapsed since now. If, for example, you want to access the
    #     Agora Service within 10 minutes after the token is generated, set expire as 600(seconds).
    # return: The room user token.
    def self.build_room_user_token(app_id, app_certificate, room_uuid, user_uuid, role, expire)
      access_token = AgoraDynamicKey2::AccessToken.new(app_id, app_certificate, expire)

      chat_user_id = Digest::MD5.hexdigest user_uuid
      service_apaas = AgoraDynamicKey2::ServiceApaas.new(room_uuid, user_uuid, role)
      service_apaas.add_privilege(AgoraDynamicKey2::ServiceApaas::PRIVILEGE_ROOM_USER, expire)
      access_token.add_service(service_apaas)

      service_rtm = AgoraDynamicKey2::ServiceRtm.new(user_uuid)
      service_rtm.add_privilege(AgoraDynamicKey2::ServiceRtm::PRIVILEGE_JOIN_LOGIN, expire)
      access_token.add_service(service_rtm)

      service_chat = AgoraDynamicKey2::ServiceChat.new(chat_user_id)
      service_chat.add_privilege(AgoraDynamicKey2::ServiceChat::PRIVILEGE_USER, expire)
      access_token.add_service(service_chat)

      access_token.build
    end

    # Build user token.
    #
    # app_id: The App ID issued to you by Agora. Apply for a new App ID from Agora Dashboard if it is missing
    #     from your kit. See Get an App ID.
    # app_certificate: Certificate of the application that you registered in the Agora Dashboard.
    #     See Get an App Certificate.
    # user_uuid: The user's id, must be unique.
    # expire: represented by the number of seconds elapsed since now. If, for example, you want to access the
    #     Agora Service within 10 minutes after the token is generated, set expire as 600(seconds).
    # return: The user token.
    def self.build_user_token(app_id, app_certificate, user_uuid, expire)
      access_token = AgoraDynamicKey2::AccessToken.new(app_id, app_certificate, expire)

      service_apaas = AgoraDynamicKey2::ServiceApaas.new('', user_uuid)
      service_apaas.add_privilege(AgoraDynamicKey2::ServiceApaas::PRIVILEGE_USER, expire)
      access_token.add_service(service_apaas)

      access_token.build
    end

    # Build app token.
    #
    # app_id: The App ID issued to you by Agora. Apply for a new App ID from Agora Dashboard if it is missing
    #     from your kit. See Get an App ID.
    # app_certificate: Certificate of the application that you registered in the Agora Dashboard.
    #     See Get an App Certificate.
    # expire: represented by the number of seconds elapsed since now. If, for example, you want to access the
    #     Agora Service within 10 minutes after the token is generated, set expire as 600(seconds).
    # return: The app token.
    def self.build_app_token(app_id, app_certificate, expire)
      access_token = AgoraDynamicKey2::AccessToken.new(app_id, app_certificate, expire)

      service_apaas = AgoraDynamicKey2::ServiceApaas.new
      service_apaas.add_privilege(AgoraDynamicKey2::ServiceApaas::PRIVILEGE_APP, expire)
      access_token.add_service(service_apaas)

      access_token.build
    end
  end
end

File: DynamicKey/AgoraDynamicKey/php/src/EducationTokenBuilder.php
<?php

require_once "AccessToken2.php";

class EducationTokenBuilder
{
    /**
     * Build user room token.
     *
     * @param $appId The App ID issued to you by Agora. Apply for a new App ID from
     *     Agora Dashboard if it is missing from your kit. See Get an App ID.
     * @param $appCertificate Certificate of the application that you registered in
     *     the Agora Dashboard. See Get an App Certificate.
     * @param $roomUuid The room's id, must be unique.
     * @param $userUuid The user's id, must be unique.
     * @param $role The user's role, such as 0(invisible), 1(teacher), 2(student), 3(assistant), 4(observer) etc.
     * @param $expire represented by the number of seconds elapsed since now. If, for example, you want to access the
     *     Agora Service within 10 minutes after the token is generated, set expireTimestamp as 600(seconds).
     * @return The user room token.
     */
    public static function buildRoomUserToken($appId, $appCertificate, $roomUuid, $userUuid, $role, $expire)
    {
        $accessToken = new AccessToken2($appId, $appCertificate, $expire);

        $chatUserId = md5($userUuid);
        $serviceEducation = new ServiceEducation($roomUuid, $userUuid, $role);
        $serviceEducation->addPrivilege($serviceEducation::PRIVILEGE_ROOM_USER, $expire);
        $accessToken->addService($serviceEducation);

        $serviceRtm = new ServiceRtm($userUuid);
        $serviceRtm->addPrivilege($serviceRtm::PRIVILEGE_LOGIN, $expire);
        $accessToken->addService($serviceRtm);

        $serviceChat = new ServiceChat($chatUserId);
        $serviceChat->addPrivilege($serviceChat::PRIVILEGE_USER, $expire);
        $accessToken->addService($serviceChat);

        return $accessToken->build();
    }

    /**
     * Build user individual token.
     *
     * @param $appId The App ID issued to you by Agora. Apply for a new App ID from
     *     Agora Dashboard if it is missing from your kit. See Get an App ID.
     * @param $appCertificate Certificate of the application that you registered in
     *     the Agora Dashboard. See Get an App Certificate.
     * @param $userUuid The user's id, must be unique.
     * @param $expire represented by the number of seconds elapsed since now. If, for example, you want to access the
     *     Agora Service within 10 minutes after the token is generated, set expireTimestamp as 600(seconds).
     * @return The build user individual token.
     */
    public static function buildUserToken($appId, $appCertificate, $userUuid, $expire)
    {
        $accessToken = new AccessToken2($appId, $appCertificate, $expire);
        $serviceEducation = new ServiceEducation("", $userUuid);
        $serviceEducation->addPrivilege($serviceEducation::PRIVILEGE_USER, $expire);
        $accessToken->addService($serviceEducation);

        return $accessToken->build();
    }

    /**
     * Build app global token.
     *
     * @param $appId The App ID issued to you by Agora. Apply for a new App ID from
     *     Agora Dashboard if it is missing from your kit. See Get an App ID.
     * @param $appCertificate Certificate of the application that you registered in
     *     the Agora Dashboard. See Get an App Certificate.
     * @param $expire represented by the number of seconds elapsed since now. If, for example, you want to access the
     *     Agora Service within 10 minutes after the token is generated, set expireTimestamp as 600(seconds).
     * @return The app global token.
     */
    public static function buildAppToken($appId, $appCertificate, $expire)
    {
        $accessToken = new AccessToken2($appId, $appCertificate, $expire);
        $serviceEducation = new ServiceEducation();
        $serviceEducation->addPrivilege($serviceEducation::PRIVILEGE_APP, $expire);
        $accessToken->addService($serviceEducation);

        return $accessToken->build();
    }
}
